### DISCUSSION TITLE
Uploading files to an external API

### DISCUSSION THREAD

**Discussion by kevincrossgrove** [vercel/next.js#49712] [Help] [answered]
### Summary

New to NextJS and trying to upload files using the following flow

Client -> NextJS Route -> External API

In the Client
```js
const formData = new FormData();
formData.set("File", file);

return fetch("/api/file/upload", {
  method: "POST",
  body: formData,
  headers: {
    "Content-Type": "multipart/form-data",
  },
})
```

In the NextJS API route

```js
return fetch(process.env.API_HOST + "tenant/file", {
    method: "POST",
    body: formData,
    headers: {
      "Content-Type": "multipart/form-data",
      Authorization: "Bearer " + process.env.API_TOKEN,
    },
  })
```

My external API is a NodeJS express server. It uses the `express-fileupload` middleware
My endpoint is returning "Please provide a file to continue"

```js
let files: any = req.files;

if (!req.files || Object.keys(req.files).length === 0) {
  return Responses.BadRequest(res, "Please provide a file to continue");
}
```

**Comment by koteus** [comment, marked as answer]
If you use `multipart/form-data` content type, you are required to specify the boundary parameter. 
You don't have to specify it manually though the browser adds boundary automatically like this: 
```
Content-Type: multipart/form-data; boundary=---WebKitFormBoundary7MA4YWxkTrZu0gW

    -----WebKitFormBoundary7MA4YWxkTrZu0gW
    Content-Disposition: form-data; name=”file”; filename=”captcha”
    Content-Type:

    -----WebKitFormBoundary7MA4YWxkTrZu0gW
    Content-Disposition: form-data; name=”action”

    submit
    -----WebKitFormBoundary7MA4YWxkTrZu0gW--
```

So, this part `WebKitFormBoundary7MA4YWxkTrZu0gW` is added by the browser. In order to make your code work you can remove `Content-Type` header from the frontend code:
```javascript
const formData = new FormData();
formData.set("File", file);

return fetch("/api/file/upload", {
  method: "POST",
  body: formData,
})
```
The browser will generate the correct form type headers.

However, this solves only the first part of the problem. Because you have NextJS api route that takes this form data and re-sends it to a remote server, you need a way to generate the similar header with the boundary on the server. Because server is not a browser, it will not be done automatically this time. One of the solutions could be to use this library `form-data` (https://github.com/form-data/form-data)

If you want to learn more about boundaries and why they are needed: https://stackoverflow.com/questions/3508338/what-is-the-boundary-in-multipart-form-data
